File: lib/cretonne/meta/gen_registers.py
"""
Generate register bank descriptions for each ISA.
"""

from __future__ import absolute_import
import srcgen

try:
    from typing import Sequence, List  # noqa
    from cdsl.isa import TargetISA  # noqa
    from cdsl.registers import RegBank, RegClass  # noqa
except ImportError:
    pass


def gen_regbank(regbank, fmt):
    # type: (RegBank, srcgen.Formatter) -> None
    """
    Emit a static data definition for regbank.
    """
    with fmt.indented('RegBank {', '},'):
        fmt.format('name: "{}",', regbank.name)
        fmt.format('first_unit: {},', regbank.first_unit)
        fmt.format('units: {},', regbank.units)
        fmt.format(
                'names: &[{}],',
                ', '.join('"{}"'.format(n) for n in regbank.names))
        fmt.format('prefix: "{}",', regbank.prefix)
        fmt.format('first_toprc: {},', regbank.toprcs[0].index)
        fmt.format('num_toprcs: {},', len(regbank.toprcs))
        fmt.format(
                'pressure_tracking: {},',
                'true' if regbank.pressure_tracking else 'false')


def gen_regbank_units(regbank, fmt):
    # type: (RegBank, srcgen.Formatter) -> None
    """
    Emit constants for all the register units in `regbank`.
    """
    for unit in range(regbank.units):
        v = unit + regbank.first_unit
        if unit < len(regbank.names):
            fmt.format("{} = {},", regbank.names[unit], v)
        else:
            fmt.format("{}{} = {},", regbank.prefix, unit, v)


def gen_regclass(rc, fmt):
    # type: (RegClass, srcgen.Formatter) -> None
    """
    Emit a static data definition for a register class.
    """
    with fmt.indented(
            'pub static {}_DATA: RegClassData = RegClassData {{'
            .format(rc.name), '};'):
        fmt.format('name: "{}",', rc.name)
        fmt.format('index: {},', rc.index)
        fmt.format('width: {},', rc.width)
        fmt.format('bank: {},', rc.bank.index)
        fmt.format('toprc: {},', rc.toprc.index)
        fmt.format('first: {},', rc.bank.first_unit + rc.start())
        fmt.format('subclasses: 0x{:x},', rc.subclass_mask())
        mask = ', '.join('0x{:08x}'.format(x) for x in rc.mask())
        fmt.format('mask: [{}],', mask)
        fmt.line('info: &INFO,')
    # Also emit a convenient reference for use by hand-written code.
    fmt.line('#[allow(dead_code)]')
    fmt.format('pub static {0}: RegClass = &{0}_DATA;', rc.name)


def gen_isa(isa, fmt):
    # type: (TargetISA, srcgen.Formatter) -> None
    """
    Generate register tables for isa.
    """
    if not isa.regbanks:
        print('cargo:warning={} has no register banks'.format(isa.name))

    with fmt.indented('pub static INFO: RegInfo = RegInfo {', '};'):
        # Bank descriptors.
        with fmt.indented('banks: &[', '],'):
            for regbank in isa.regbanks:
                gen_regbank(regbank, fmt)
        with fmt.indented('classes: &[', '],'):
            for rc in isa.regclasses:
                fmt.format('&{}_DATA,', rc.name)

    # Register class descriptors.
    for rc in isa.regclasses:
        gen_regclass(rc, fmt)

    # Emit constants for all the register units.
    fmt.line('#[allow(dead_code, non_camel_case_types)]')
    fmt.line('#[derive(Clone, Copy)]')
    with fmt.indented('pub enum RU {', '}'):
        for regbank in isa.regbanks:
            gen_regbank_units(regbank, fmt)
    with fmt.indented('impl Into<RegUnit> for RU {', '}'):
        with fmt.indented('fn into(self) -> RegUnit {', '}'):
            fmt.line('self as RegUnit')


def generate(isas, out_dir):
    # type: (Sequence[TargetISA], str) -> None
    for isa in isas:
        fmt = srcgen.Formatter()
        gen_isa(isa, fmt)
        fmt.update_file('registers-{}.rs'.format(isa.name), out_dir)

File: lib/codegen/meta/src/gen_registers.rs
use cdsl::isa::TargetIsa;
use cdsl::regs::{RegBank, RegClass};
use cranelift_entity::EntityRef;
use error;
use srcgen::Formatter;

fn gen_regbank(fmt: &mut Formatter, reg_bank: &RegBank) {
    let names = if reg_bank.names.len() > 0 {
        format!(r#""{}""#, reg_bank.names.join(r#"", ""#))
    } else {
        "".to_string()
    };
    fmt.line("RegBank {");
    fmt.indent(|fmt| {
        fmt.line(&format!(r#"name: "{}","#, reg_bank.name));
        fmt.line(&format!("first_unit: {},", reg_bank.first_unit));
        fmt.line(&format!("units: {},", reg_bank.units));
        fmt.line(&format!("names: &[{}],", names));
        fmt.line(&format!(r#"prefix: "{}","#, reg_bank.prefix));
        fmt.line(&format!("first_toprc: {},", reg_bank.toprcs[0].index()));
        fmt.line(&format!("num_toprcs: {},", reg_bank.toprcs.len()));
        fmt.line(&format!(
            "pressure_tracking: {},",
            if reg_bank.pressure_tracking {
                "true"
            } else {
                "false"
            }
        ));
    });
    fmt.line("},");
}

fn gen_regclass(isa: &TargetIsa, reg_class: &RegClass, fmt: &mut Formatter) {
    let reg_bank = isa.reg_banks.get(reg_class.bank).unwrap();

    let mask: Vec<String> = reg_class
        .mask(reg_bank.first_unit)
        .iter()
        .map(|x| format!("0x{:08x}", x))
        .collect();
    let mask = mask.join(", ");

    fmt.line(&format!(
        "pub static {}_DATA: RegClassData = RegClassData {{",
        reg_class.name
    ));
    fmt.indent(|fmt| {
        fmt.line(&format!(r#"name: "{}","#, reg_class.name));
        fmt.line(&format!("index: {},", reg_class.index.index()));
        fmt.line(&format!("width: {},", reg_class.width));
        fmt.line(&format!("bank: {},", reg_class.bank.index()));
        fmt.line(&format!("toprc: {},", reg_class.toprc.index()));
        fmt.line(&format!(
            "first: {},",
            reg_bank.first_unit + reg_class.start
        ));
        fmt.line(&format!("subclasses: {:#x},", reg_class.subclass_mask()));
        fmt.line(&format!("mask: [{}],", mask));
        fmt.line("info: &INFO,");
    });
    fmt.line("};");
    fmt.line("#[allow(dead_code)]");
    fmt.line(&format!(
        "pub static {}: RegClass = &{}_DATA;",
        reg_class.name, reg_class.name
    ));
}

fn gen_regbank_units(reg_bank: &RegBank, fmt: &mut Formatter) {
    for unit in 0..reg_bank.units {
        let v = unit + reg_bank.first_unit;
        if (unit as usize) < reg_bank.names.len() {
            fmt.line(&format!("{} = {},", reg_bank.names[unit as usize], v));
            continue;
        }
        fmt.line(&format!("{}{} = {},", reg_bank.prefix, unit, v));
    }
}

fn gen_isa(isa: &TargetIsa, fmt: &mut Formatter) -> Result<(), error::Error> {
    // Emit RegInfo.
    fmt.line("pub static INFO: RegInfo = RegInfo {");

    fmt.indent(|fmt| {
        fmt.line("banks: &[");
        // Bank descriptors.
        fmt.indent(|fmt| {
            for reg_bank in isa.reg_banks.values() {
                gen_regbank(fmt, &reg_bank);
            }
        });
        fmt.line("],");
        // References to register classes.
        fmt.line("classes: &[");
        fmt.indent(|fmt| {
            for reg_class in isa.reg_classes.values() {
                fmt.line(&format!("&{}_DATA,", reg_class.name));
            }
        });
        fmt.line("],");
    });
    fmt.line("};");

    // Register class descriptors.
    for rc in isa.reg_classes.values() {
        gen_regclass(&isa, rc, fmt);
    }

    // Emit constants for all the register units.
    fmt.line("#[allow(dead_code, non_camel_case_types)]");
    fmt.line("#[derive(Clone, Copy)]");
    fmt.line("pub enum RU {");
    fmt.indent(|fmt| {
        for reg_bank in isa.reg_banks.values() {
            gen_regbank_units(reg_bank, fmt);
        }
    });
    fmt.line("}");

    // Emit Into conversion for the RU class.
    fmt.line("impl Into<RegUnit> for RU {");
    fmt.indent(|fmt| {
        fmt.line("fn into(self) -> RegUnit {");
        fmt.indent(|fmt| {
            fmt.line("self as RegUnit");
        });
        fmt.line("}")
    });
    fmt.line("}");

    Ok(())
}

pub fn generate(isa: TargetIsa, base_filename: &str, out_dir: &str) -> Result<(), error::Error> {
    let mut fmt = Formatter::new();
    gen_isa(&isa, &mut fmt)?;
    fmt.update_file(&format!("{}-{}.rs", base_filename, isa.name), out_dir)?;
    Ok(())
}
